Implement swagger for KEB

polskikiel opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow contributing guidelines, templates, the recommended Git workflow, and any related documentation.
2. Read and submit the required Contributor Licence Agreements (https://github.com/kyma-project/community/blob/master/CONTRIBUTING.md#agreements-and-licenses).
3. Test your changes and attach their results to the pull request.
4. Update the relevant documentation.
-->

**Description**

Changes proposed in this pull request:

- Implemented swagger for KEB

**Related issue(s)**
<!-- If you refer to a particular issue, provide its number. For example, `Resolves #123`, `Fixes #43`, or `See also #33`. -->

polskikiel commented:
/retest

piotrmiskiewicz commented:
Can you add the swagger console to the root?

piotrmiskiewicz commented:
Add description and possible values, for example "target":"all", most of docs we can find in the code: ```// RuntimeTarget captures a specification of SKR targets to resolve for an orchestration.
// When a RuntimeTarget defines multiple fields, all should match to any given runtime to be selected (i.e. the terms are AND-ed).
type RuntimeTarget struct {
	// Valid values: "all"
	Target string `json:"target,omitempty"`
	// Regex pattern to match against the runtime's GlobalAccount field. E.g. CA50125541TID000000000741207136, CA.*
	GlobalAccount string `json:"globalAccount,omitempty"`
	// Regex pattern to match against the runtime's SubAccount field. E.g. 0d20e315-d0b4-48a2-9512-49bc8eb03cd1
	SubAccount string `json:"subAccount,omitempty"`
	// Regex pattern to match against the shoot cluster's Region field (not SCP platform-region). E.g. "europe|eu-"
	Region string `json:"region,omitempty"`
	// RuntimeID is used to indicate a specific runtime
	RuntimeID string `json:"runtimeID,omitempty"`
	// PlanName is used to match runtimes with the same plan
	PlanName string `json:"planName,omitempty"`
}```

piotrmiskiewicz commented on docs/kyma-environment-broker/08-07-swagger.md:
```diff
@@ -0,0 +1,39 @@
+
+This tutorial shows how to check the API schema using Swagger.
+
+The Swagger is injected into the application and it's exposes the Swagger UI on the `/swagger/` endpoint.
```
the endpoint has been changed

piotrmiskiewicz commented:
1. can the schema define enums? for example for target all
2. runtimes endpoint does not work for me (I'm port-frowarding)

polskikiel commented:
@piotrmiskiewicz regarding 1, it's possible so I added enums wherever it was doable. About the second thing, it works now, after I moved it after `runtime.AttachRoutes()` execution at the end of `main.go` file

klaudiagrz commented on docs/kyma-environment-broker/08-07-swagger.md:
```diff
@@ -0,0 +1,39 @@
+---
+title: Check API using swagger
+type: Tutorials
```
I'm not sure Tutorials is the right type here since it's not really a step-by-step instruction 🤔 I'd rather classify this as a Details doc.

klaudiagrz commented on docs/kyma-environment-broker/08-07-swagger.md:
```diff
@@ -0,0 +1,39 @@
+
+This tutorial shows how to check the API schema using Swagger.
+
+The Swagger is injected into the KEB's container which exposes the Swagger UI on root endpoint.
+
+The Swagger UI static files are copied from the [official source](https://github.com/swagger-api/swagger-ui/tree/master/dist).
+
+The KEB uses swagger schema file mounted as volume to the Pod. You can find that schema [here](https://github.com/kyma-project/control-plane/blob/master/resources/kcp/charts/kyma-environment-broker/files/swagger.yaml).
```
These sentences seem to be a bit out of context 🤔 Let's provide some general context to them.

klaudiagrz commented on docs/kyma-environment-broker/08-07-swagger.md:
````diff
@@ -0,0 +1,39 @@
+   https://$BROKER_URL/
+   ```
+
+> **NOTE:** Using this option you can't use `Try it out` option due to not configured OAuth2 swagger schema.
````
```suggestion
> **NOTE:** If you choose this option, you can't use the `Try it out` feature as the OAuth2 Swagger schema is not configured.
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        globalAccount:
+          type: string
+          example: 054ac2c2-318f-45dd-855c-eee41513d40d
+          description: Regex pattern to match against the runtime's GlobalAccount field
```
```suggestion
          description: Regex pattern to match against Runtime's globalAccount field
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        subAccount:
+          type: string
+          example: 054ac2c2-318f-45dd-855c-eee41513d40d
+          description: Regex pattern to match against the runtime's SubAccount field
```
```suggestion
          description: Regex pattern to match against Runtime's subAccount field
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        region:
+          type: string
+          example: europe|eu-
+          description: Regex pattern to match against the shoot cluster's Region field (not SCP platform-region)
```
```suggestion
          description: Regex pattern to match against the Shoot cluster's region field (not SCP platform region)
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        shootName:
+          type: string
+          example: c-8e9ea4f
+          description: Name of the shoot on gardener
```
```suggestion
          description: Name of the Shoot cluster on Gardener
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        maintenanceWindowBegin:
+          type: string
+          example: 150405-0000
+          description: Start time of orchestrations processing
```
```suggestion
          description: Start time of the orchestrations processing
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        dryRun:
+          type: boolean
+          default: false
+          description: Specifies if orchestration is for testing purposes
```
```suggestion
          description: Specifies if the orchestration is used for testing purposes
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+          type: string
+          format: uuid
+          example: 054ac2c2-318f-45dd-855c-eee41513d40d
+          description: ID of the service plan
```
```suggestion
          description: ID of the service plan
```

polskikiel replied:
glad we agree

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        servicePlanName:
+          type: string
+          example: azure
+          description: Specifies plan name
```
```suggestion
          description: Specifies the plan name
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        shootName:
+          type: string
+          example: c-8e9ea4f
+          description: Name of the shoot on gardener
```
```suggestion
          description: Name of the Shoot cluster on Gardener
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        maintenanceWindowBegin:
+          type: string
+          example: 150405-0000
+          description: Start time of orchestrations processing
```
```suggestion
          description: Start time of the orchestrations processing
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        dryRun:
+          type: boolean
+          default: false
+          description: Specifies if orchestration is for testing purposes
```
```suggestion
          description: Specifies if the orchestration is used for testing purposes
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        servicePlanName:
+          type: string
+          example: azure
+          description: Specifies plan name
```
```suggestion
          description: Specifies the plan name
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+          description: Specifies plan name
+        kymaConfig:
+          type: string
+          description: Object with kyma config send to provisioner
```
```suggestion
          description: Object with the Kyma config sent to Runtime Provisioner
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+          description: Object with kyma config send to provisioner
+        clusterConfig:
+          type: string
+          description: Object with cluster config send to provisioner
```
```suggestion
          description: Object with the cluster config sent to Runtime Provisioner
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+        shootName:
+          type: string
+          example: c-8e9ea4f
+          description: Name of the shoot on gardener
```
```suggestion
          description: Name of the Shoot cluster on Gardener
```

klaudiagrz commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+      properties:
+        error:
+          type: string
+          example: "while decoding request body: invalid character '}' looking for beginning of object key string"
```
```suggestion
          example: "Error when decoding the request body: invalid character '}' looking for beginning of object key string"
```

polskikiel commented on resources/kcp/charts/kyma-environment-broker/files/swagger.yaml:
```diff
@@ -0,0 +1,576 @@
+    post:
+      summary: orchestrates kyma upgrade
+      operationId: upgradeKyma
+      description: Start the processing of Kyma upgrade, returns orchestration ID
```
not possible :(

polskikiel commented:
/retest

klaudiagrz commented on docs/kyma-environment-broker/03-11-swagger.md:
```diff
@@ -0,0 +1,36 @@
+type: Details
+---
+
+Using the Swagger UI you can visualize the KEB's APIs on a single page.
```
```suggestion
Using the Swagger UI, you can visualize Kyma Environment Broker's (KEB's) APIs on a single page.
```

klaudiagrz commented on docs/kyma-environment-broker/03-11-swagger.md:
```diff
@@ -0,0 +1,36 @@
+
+Using the Swagger UI you can visualize the KEB's APIs on a single page.
+
+The Swagger UI static files are copied from the [official source](https://github.com/swagger-api/swagger-ui/tree/master/dist) then they're injected into the KEB's container which exposes them on the root endpoint.
```
```suggestion
The Swagger UI static files are copied from the [official source](https://github.com/swagger-api/swagger-ui/tree/master/dist) and then they are injected into KEB's container which exposes them on the root endpoint.
```

klaudiagrz commented on docs/kyma-environment-broker/03-11-swagger.md:
```diff
@@ -0,0 +1,36 @@
+
+The KEB uses swagger schema file mounted as volume to the Pod. You can find that schema [here](https://github.com/kyma-project/control-plane/blob/master/resources/kcp/charts/kyma-environment-broker/files/swagger.yaml).
+
+You can use several ways to expose and use the Swagger UI, but two of them are recommended.
```
```suggestion
You can either use Virtual Service or port-forward the Pod to expose and use the Swagger UI.
```

klaudiagrz commented on docs/kyma-environment-broker/03-11-swagger.md:
```diff
@@ -0,0 +1,36 @@
+
+The Swagger UI static files are copied from the [official source](https://github.com/swagger-api/swagger-ui/tree/master/dist) then they're injected into the KEB's container which exposes them on the root endpoint.
+
+The KEB uses swagger schema file mounted as volume to the Pod. You can find that schema [here](https://github.com/kyma-project/control-plane/blob/master/resources/kcp/charts/kyma-environment-broker/files/swagger.yaml).
```
```suggestion
KEB uses a Swagger schema file that is mounted as a volume to the Pod. You can find that schema [here](https://github.com/kyma-project/control-plane/blob/master/resources/kcp/charts/kyma-environment-broker/files/swagger.yaml).
```

polskikiel commented:
/test pre-master-control-plane-gke-integration
